TLSCertificateManagement Cluster

gmarcosb opened the pull request:
Split out from https://github.com/project-chip/connectedhomeip/pull/35773:

Add camera TLS clusters from https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/10004

Generated with:
```
 ./alchemy zap --attribute="in-progress"  --sdkRoot=[] --specRoot=[]  'TLSCertificateManagement.adoc' 
 ./scripts/tools/zap_regen_all.py
```

#### Testing
Verified by CI

andy31415 commented on src/app/zap-templates/zcl/zcl-with-test-extensions.json:
```diff
@@ -291,8 +292,7 @@
             "RegisteredClients",
             "ICDCounter",
-            "ClientsSupportedPerFabric",
-            "MaximumCheckInBackOff"
```
What does MaximumCheckInBackOff have to do with TLSCertificateManagement?

gmarcosb replied:
It looks like someone manually modified this instead of using Alchemy, resulting in an invalid state; `MaximumCheckInBackOff` is tagged as `apiMaturity="provisional"` so it should not be here

gmarcosb replied:
I'm going to dig in to whether this was an issue with the spec or with Alchemy (or user error on my part w/ unclean client), manually fixed

andy31415 commented on src/app/zap-templates/zcl/data-model/chip/tls-certificate-management-cluster.xml:
```diff
@@ -0,0 +1,142 @@
+-->
+<!--
+XML generated by Alchemy; DO NOT EDIT.
+Source: src/tls/TLSCertificateManagement.adoc
```
I cannot find the `tls` directory in https://github.com/CHIP-Specifications/connectedhomeip-spec/tree/master/src

Should we wait for the spec to at least have a provisional item for this? What is the development order here? I am concerned that we may merge things that will become different from the spec.

gmarcosb replied:
You have to look at the camera branch, linked from PR: https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/10004

gmarcosb commented on src/app/zap-templates/zcl/data-model/chip/tls-certificate-management-cluster.xml:
```diff
@@ -0,0 +1,142 @@
+    <item fieldId="2" name="IntermediateCerts" array="true" type="octet_string"/>
+  </struct>
+
+  <cluster>
```
Woops, good looking out!

bzbarsky-apple commented on src/app/zap-templates/zcl/zcl.json:
```diff
@@ -285,8 +286,7 @@
             "ICDCounter",
-            "ClientsSupportedPerFabric",
-            "MaximumCheckInBackOff"
+            "ClientsSupportedPerFabric"
```
This change is wrong.  MaximumCheckInBackOff is in fact implemented in the relevant cluster, and is implemented via AttributeAccessInterface.

If the tool made this change, the tool needs to be fixed.

gmarcosb replied:
I'm not sure what's going on, if it's user error on my part where I have some stale file messing things up (Hasty was unable to repro), or if it's an issue with Alchemy

I've manually updated the files properly & will do some Alchemy debugging to see what's up

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/tls-certificate-management-cluster.xml:
```diff
@@ -0,0 +1,142 @@
+  <struct name="TLSClientCertificateDetailStruct" apiMaturity="provisional">
+    <cluster code="0x0801"/>
+    <item fieldId="0" name="CCDID" type="int16u"/>
+    <item fieldId="1" name="ClientCertificate" type="long_octet_string" length="3000"/>
+    <item fieldId="2" name="IntermediateCerts" array="true" type="octet_string"/>
```
Spec comment: why is "Certificate" abbreviated in one field name but not the other?  Seems like a weird inconsistency.  Please make sure a spec issue is filed.

aronrosenberg replied:
Will rename to IntermediateCertificates in spec.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/tls-certificate-management-cluster.xml:
```diff
@@ -0,0 +1,142 @@
+      <description>This command SHALL provision the provided certificate for the passed in CAID.</description>
+      <access op="invoke" privilege="administer"/>
+      <arg id="0" name="Certificate" type="long_octet_string" length="3000"/>
+      <arg id="1" name="CAID" type="int16u" optional="true" isNullable="true"/>
```
Spec issue: Why is this optional and nullable?  What are the semantics of "present but null", and how do those differ from "absent", if at all?  And if they don't differ, why this setup?

aronrosenberg replied:
Will change to mandatory and adjust language to indicate null's meaning as install new cert.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/tls-certificate-management-cluster.xml:
```diff
@@ -0,0 +1,142 @@
+
+    <command code="0x02" source="client" name="FindRootCertificate" optional="false" response="FindRootCertificateResponse">
+      <description>This command SHALL return the TLSCertStruct for the passed in CAID.</description>
+      <arg id="0" name="CAID" type="int16u" optional="true" isNullable="true"/>
```
Spec issue: why is this field optional?

aronrosenberg replied:
Will change to mandatory and adjust language to indicate null's meaning.

gmarcosb committed:
Generated using:
./alchemy zap --attribute="in-progress" --sdkRoot=[] --specRoot=[] 'TLSCertificateManagement.adoc'

With manual edits adding description & apiMaturity

gmarcosb committed:
Generated using ./scripts/tools/zap_regen_all.py

gmarcosb committed:
Merge branch 'master' into camera-webrtc-server
